File: Authenticator-ObjC/EHFAuthenticator.m
//
//  EHFAuthenticator.m
//  EHFoundation
//
//  Created by Christopher Truman on 7/31/14.
//  Copyright (c) 2014 eHarmony. All rights reserved.
//

#import "EHFAuthenticator.h"

@interface EHFAuthenticator ()

@property (nonatomic, strong) LAContext * context;

@end

@implementation EHFAuthenticator

+ (BOOL) canAuthenticateWithError:(NSError **) error
{
    if ([NSClassFromString(@"LAContext") class]) {
        if ([[EHFAuthenticator sharedInstance].context canEvaluatePolicy:[EHFAuthenticator sharedInstance].policy error:error]) {
            return YES;
        }
        return NO;
    }
    return NO;
}

static EHFAuthenticator *sharedInstance;

+ (EHFAuthenticator *) sharedInstance
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[EHFAuthenticator alloc] init];
    });
    
    return sharedInstance;
}

- (instancetype)init{
    if (self = [super init]) {
        self.context = [[LAContext alloc] init];
        self.useDefaultFallbackTitle = NO;
        self.hideFallbackButton = NO;
        self.policy = LAPolicyDeviceOwnerAuthenticationWithBiometrics;
    }
    return self;
}

- (void) authenticateWithSuccess:(EHFCompletionBlock) success andFailure:(EHFAuthenticationErrorBlock) failure
{
    self.context = [[LAContext alloc] init];
    NSError *authError = nil;
    if (self.useDefaultFallbackTitle) {
        self.context.localizedFallbackTitle = self.fallbackButtonTitle;
    }else if (self.hideFallbackButton){
        self.context.localizedFallbackTitle = @"";
    }
    if ([self.context canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&authError]) {
        [self.context evaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics
                     localizedReason:self.reason
                               reply:^(BOOL authenticated, NSError *error) {
                                   if (authenticated) {
                                       DispatchMainThread(^(){success();});
                                   } else {
                                       DispatchMainThread(^(){failure((LAError) error.code);});
                                   }
                               }];
    } else {
        failure((LAError) authError.code);
    }
}

@end

File: Authenticator-Swift/EHFAuthenticator.swift
//
//  EHFAuthenticator.swift
//  EHFAuthenticator-Example
//
//  Created by Christopher Truman on 9/29/14.
//  Copyright (c) 2014 truman. All rights reserved.
//

import Foundation
import LocalAuthentication

typealias EHFCompletionBlock = (Void) ->()
typealias EHFAuthenticationErrorBlock = (Int) -> ()

class EHFAuthenticator : NSObject {
    
    fileprivate var context : LAContext
    
    // reason string presented to the user in auth dialog
    var reason : NSString
    
    // Allows fallback button title customization. If set to nil, "Enter Password" is used.
    var fallbackButtonTitle : NSString
    
    // If set to NO it will not customize the fallback title, shows default "Enter Password".  If set to YES, title is customizable.  Default value is NO.
    var useDefaultFallbackTitle : Bool
    
    // Disable "Enter Password" fallback button. Default value is NO.
    var hideFallbackButton : Bool
    
    // Default value is LAPolicyDeviceOwnerAuthenticationWithBiometrics.  This value will be useful if LocalAuthentication.framework introduces new auth policies in future version of iOS.
    var policy : LAPolicy
    
    
    class var sharedInstance : EHFAuthenticator {
        struct Static {
            static let instance : EHFAuthenticator = EHFAuthenticator()
        }
        return Static.instance
    }
    
    override init(){
        self.context = LAContext()
        self.fallbackButtonTitle = ""
        self.useDefaultFallbackTitle = false
        self.hideFallbackButton = false
        self.policy = .deviceOwnerAuthenticationWithBiometrics
        self.reason = ""
    }
    
    class func canAuthenticateWithError() throws{
        let error: NSError! = NSError(domain: "Migrator", code: 0, userInfo: nil)
        if ((NSClassFromString("LAContext")) != nil){
            if (EHFAuthenticator.sharedInstance.context.canEvaluatePolicy(EHFAuthenticator.sharedInstance.policy, error: nil)){
                return
            }
            throw error
        }
        throw error
    }
    
    func authenticateWithSuccess(_ success: @escaping EHFCompletionBlock, failure: @escaping EHFAuthenticationErrorBlock){
        self.context = LAContext()
        if (self.useDefaultFallbackTitle) {
            self.context.localizedFallbackTitle = self.fallbackButtonTitle as String;
        }else if (self.hideFallbackButton){
            self.context.localizedFallbackTitle = "";
        }
        if (self.context.canEvaluatePolicy(policy, error: nil)) {
            self.context.evaluatePolicy(policy, localizedReason:
                reason as String, reply:{ authenticated, error in
                if (authenticated) {
                    DispatchQueue.main.async(execute: {success()})
                } else {
                    DispatchQueue.main.async(execute: {failure(error!._code)})
                }
            })
        } else {
            failure(0)
        }
    }
}
